designs/component-config.md

```diff
@@ -0,0 +1,241 @@
+
+```golang
+package config
+
+import (
+  "time"
+) 
+
+func (in *ControllerNameConfiguration) GetSyncPeriod() *time.Duration {}
+func (in *ControllerNameConfiguration) GetLeaderElection() bool {}
+func (in *ControllerNameConfiguration) GetLeaderElectionNamespace() string {}
+func (in *ControllerNameConfiguration) GetLeaderElectionID() string {}
+
+func (in *ControllerNameConfiguration) GetLeaseDuration() *time.Duration {}
+func (in *ControllerNameConfiguration) GetRenewDeadline() *time.Duration {}
+func (in *ControllerNameConfiguration) GetRetryPeriod() *time.Duration {}
+
+func (in *ControllerNameConfiguration) GetNamespace() string {}
+func (in *ControllerNameConfiguration) GetMetricsBindAddress() string {}
+func (in *ControllerNameConfiguration) GetHealthProbeBindAddress() string {}
+
+func (in *ControllerNameConfiguration) GetReadinessEndpointName() string {}
+func (in *ControllerNameConfiguration) GetLivenessEndpointName() string {}
+
+func (in *ControllerNameConfiguration) GetPort() int {}
+func (in *ControllerNameConfiguration) GetHost() string {}
+
+func (in *ControllerNameConfiguration) GetCertDir() string {}
+```
+
+Besides the implementation of the `ComponentConfig` The controller author as it stands would also need to implement the unmarshalling of the `ConfigMap` into the `ComponentConfig`, for this `controller-runtime` could expose helper methods to load a file from disk, unmarshal to the struct and pass the pointer into the `NewFromComponentConfig()` to return the `ctrl.Manager`
+
+### User Stories
+
+#### Controller Author with `controller-runtime`
+
+- Implement `ComponentConfig` type
+- Implement `ManagerConfiguration` interface for `ComponentConfig` object
+- Set up `ConfigMap` unmarshalling into `ComponentConfig` type
+- Initialize `ctrl.Manager` with `NewFromComponentConfig`
+- Build custom controller as usual
+
+#### Controller Author with `kubebuilder` (tbd proposal for `kubebuilder`)
+
+- Initialize `kubebuilder` project using `--component-config-name=XYZConfiguration`
+- Build custom controller as usual
+
+#### Controller User without modifications to config
+
+_Provided that the controller provides manifests_
+
+- Apply the controller to the cluster
+- Deploy custom resources
+
+#### Controller User with modifications to config
+
+- _Following from previous example without changes_
+- Create a new `ConfigMap` for changes
+- Modify the `controller-runtime` pod to use the new `ConfigMap`
+- Apply the controller to the cluster
+- Deploy custom resources
+
+
+### Risks and Mitigations
+
+- Given that this isn't changing the core Manager initialization for `controller-runtime` it's fairly low risk
+- If the underlaying `ctrl.Options{}`
+
+## Alternatives
+
+* `NewFromComponentConfig()` could load the object from disk and hydrate the `ComponentConfig` type
```

Comment:
Would the `<ComponentConfig>{}` then be used directly within the manager code, thus deprecating `ctrl.Options{}` for this alternative solution?

Reply:
Good question, originally when I wrote this I was thinking that `NewFromComponentConfig()` should also be responsible for parsing out the config file but when I started to implement this as a demo I found it slightly difficult and would require `flags` to also be standardized which is something `controller-runtime` shouldn't care about. 

Here is an example implementation - https://github.com/christopherhein/github-controller/pull/5/files#diff-7ddfb3e035b42cd70649cc33393fe32cR72-R79

Which uses - https://github.com/christopherhein/controller-runtime/commit/c3954c0a2f5a#diff-7d903b129bc883285ffc5068c39c00b0R392-R403

I'm still debating if the file parsing should be done in `controller-runtime` especially since this example doesn't support flag overrides, I would love to see these functions make their way into https://git.k8s.io/component-base

Reply:
Regarding deprecating `ctrl.Options{}` for backwards compatibility reasons I'm concerned about trying to tackle that in the same proposal, it's currently what is used to set up any `ctrl.NewManager()` which is scaffolded w/ `kubebuilder`

Reply:
Yeah, great idea @shawn-hurley let me write something up and we can discuss here.

Reply:
The nice thing I think we can leverage is the `consts` that are defined for `NewManager` which would just require us to change what that references. As of now it's `manager.New()` path wise we'd be able to leverage that and change it's references to `manager.NewFromComponentConfig()`.